### ISSUE TITLE
[refine/tokenzier] 重构tokenizer接口

### ISSUE THREAD

**Issue by Mddct** [wenet-e2e/wenet#2160]
###  相关pr 提到的
- https://github.com/wenet-e2e/wenet/issues/2142#issuecomment-1823782110
- https://github.com/wenet-e2e/wenet/issues/2142
- https://github.com/wenet-e2e/wenet/pull/2159

###  ref:
- https://github.com/espnet/espnet/tree/master/espnet2/text
- https://huggingface.co/docs/transformers/main_classes/tokenizer
- https://github.com/openai/whisper/blob/main/whisper/tokenizer.py

### 涉及到的文件：
-  wenet/dataset/dataset.py
-  wenet/dataset/process.py
-  wenet/bin/train.py
-  wenet/bin/recognize.py
-  wenet/utils/tokenize_utils.py
- wenet/utils/context_graph.py
- wenet/bin/aligment.py
- wene/tools/text2token.py
- wenet/test/test_tokenize.py
- wenet/test/whisper/test_whisper.py
- wenet/cli/model.py

TODO:
- [x] tokenzier interface
- [x] bpe 
- [x] paraformer
- [x] char
- [x] unit test

**Comment by Mddct** [comment]
I write a  code snippet like below， any suggestion？ @xingchensong  @robin1001 bin
```python
from abc import ABC, abstractmethod
from collections.abc import Iterable
from typing import List, Tuple


class BaseTokenizer(ABC):

    def tokenize(self, line: str) -> Tuple[List[str], List[int]]:
        tokens = self.text2tokens(line)
        ids = self.tokens2ids(tokens)
        return tokens, ids

    def detaokenzie(self, ids: List[int]) -> Tuple[str, List[str]]:
        tokens = self.ids2tokens(ids)
        text = self.tokens2text(tokens)
        return text, tokens

    @abstractmethod
    def text2tokens(self, line: str) -> List[str]:
        raise NotImplementedError("abstract method")

    @abstractmethod
    def tokens2text(self, tokens: Iterable[str]) -> str:
        raise NotImplementedError("abstract method")

    @abstractmethod
    def tokens2ids(self, tokens: List[str]) -> List[int]:
        raise NotImplemented("abstract method")

    @abstractmethod
    def ids2tokens(self, ids: List[int]) -> List[str]:
        raise NotImplemented("abstract method")
```

```python
import re

from collections.abc import Iterable
from os import PathLike
from typing import List, Optional
from wenet.utils.file_utils import read_symbol_table, read_non_lang_symbols
from wenet.utils.text.base_tokenzier import BaseTokenizer
from wenet.utils.text.tokenize_utils import tokenize_by_bpe_model


class WenetTokenizer(BaseTokenizer):
    """Wrapper for original wenet tokenize implementation
    """

    def __init__(
        self,
        symbol_table: PathLike,
        bpe_model: Optional[PathLike] = None,
        non_lang_syms: Optional[PathLike] = None,
        split_with_space: bool = False,
        connect_symbol: str = '',
    ) -> None:
        self.non_lang_syms_pattern = None
        if non_lang_syms is not None:
            self.non_lang_syms_pattern = re.compile(
                r"(\[[^\[\]]+\]|<[^<>]+>|{[^{}]+})")

        self.symbol_table = read_symbol_table(symbol_table)
        self.non_lang_syms = read_non_lang_symbols(non_lang_syms)
        self.bpe_model = None
        if bpe_model is not None:
            import sentencepiece as spm
            self.bpe_model = spm.SentencePieceProcessor()
            self.bpe_model.load(bpe_model)
        self.char_dict = {v: k for k, v in self.symbol_table.items()}
        self.split_with_space = split_with_space
        self.connect_symbol = connect_symbol

    def text2tokens(self, line: str) -> List[str]:
        if self.non_lang_syms_pattern is not None:
            parts = self.non_lang_syms_pattern.split(line.upper())
            parts = [w for w in parts if len(w.strip()) > 0]
        else:
            parts = [line]

        tokens = []
        for part in parts:
            if part in self.non_lang_syms:
                tokens.append(part)
            else:
                if self.bpe_model is not None:
                    tokens.extend(tokenize_by_bpe_model(self.bpe_model, part))
                else:
                    if self.split_with_space:
                        part = part.split(" ")
                    for ch in part:
                        if ch == ' ':
                            ch = "▁"
                        tokens.append(ch)
        return tokens

    def tokens2text(self, tokens: Iterable[str]) -> str:
        return self.connect_symbol.join(tokens)

    def tokens2ids(self, tokens: List[str]) -> List[int]:
        ids = []
        for ch in tokens:
            if ch in self.symbol_table:
                ids.append(self.symbol_table[ch])
            elif '<unk>' in self.symbol_table:
                ids.append(self.symbol_table['<unk>'])
        return ids

    def ids2tokens(self, ids: List[int]) -> List[str]:
        content = [self.char_dict[w] for w in ids]
        return content
```

```python
class WhisperTokenizer(BaseTokenizer):
....
```

**Comment by xingchensong** [comment]
great ， i like it ♥ bpe的逻辑和char的逻辑要是能分开成两个各自的tokenozer就更好了

**Comment by robin1001** [comment]
Looks great, I like it too. 后面我们可以把针对 tokenizer 的单元测试也考虑进来。

**Comment by robin1001** [comment]
我看 proposal 中已经考虑了，Great！

**Comment by Mddct** [comment]
写BPETokenizer的时候直接把”▁“ 替换掉‘’

相关关 bpe 的 recipe 的 ： sed -e "s/▁/ /g" 的处理； 也能保持结果不变

**Comment by Mddct** [comment]
openai tiktoken 和其他实现 都会有个encode_batch, 比如
 https://github.com/openai/tiktoken/blob/39f29cecdb6fc38d9a3434e5dd15e4de58cf3c80/tiktoken/core.py#L145

但是这些实现逗使用thread pool 进行并发的

考虑到咱们后边会对dataset进行重构，以及保证tokenizer interface 简介，咱们可以不需要tokenize_batch这些功能

**Comment by robin1001** [comment]
Yes，Compared with WenetTokenizer, CharTokenizer and BpeTokenizer are more clear.

**Comment by xingchensong** [comment]
还有个需求需要考虑下，一个tokenizer允许存在俩个不同的子tokenizer。比如这个场景，ctc的词表和decoder的词表不共享，那么ctc需要一个tokenizer（如char tokenizer），decoder需要一个tokenizer（如whisper tokenizer）

**Comment by Mddct** [comment]
> 还有个需求需要考虑下，一个tokenizer允许存在俩个不同的子tokenizer。比如这个场景，ctc的词表和decoder的词表不共享，那么ctc需要一个tokenizer（如char tokenizer），decoder需要一个tokenizer（如whisper tokenizer）

old dataset io
``python

whisper_tokenizer = WhisperTokenizer(...)
ctc_tokenizer = BpeTokenizer | CustomTokenizer

def multi_tokenize(sample, *tokenizers)：
     for data in sample：
        for tokenizer in tokenizers:
            name = tokenizer.name
            tokens， label = tokenizer.tokenize(data[’txt’])
            data[name]['label'] = label
            data[name]['tokens'] = tokens
        yield data
            
```

new dataset io
```
whisper_tokenizer = WhisperTokenizer(...)
ctc_tokenizer = BpeTokenizer | CustomTokenizer

wenet_dataset = get_dataset (...)

whisper_dataset = wenet_dataset.parallel_map(whisper_tokenizer)
ctc_dataset = wenet_dataset.parallel_map(ctc_tokenizer)

# 这里我们就有了针对不同任务的tokens和labels 
wenet_dataset = dataset.zip(whisper_dataset, ctc_dataset)

```

**Comment by Mddct** [comment]
或者方案2, 把这个逻辑放到单独的一个类里边:

```python
class MultiNamedTokenizer:
    def __init__(*tokenizers):
        self.tokenizers = []
        for tokenizer in tokenizers:
           self.tokenizers.append(tokenizer)

    def tokenize(line :str):
       labels = {}
       tokens = {}
        for tokenizer in self.tokenizers:
            name = tokenizer.name
            tokens, label = tokenizer.tokenize(line)
            labels[name] = labels
            tokens[name] = tokens
       return labels, tokens

ctc_tokenizer = CharToeknzer(..., name='ctc')
whisper_tokenizer = WhisperTokenizer(...., name='whisper')

task_tokenizer = MultiNamedTokenizer(ctc_tokenizer, whisper_tokenizer)

# old io
def named_tokenizer(sample, tokenizer)：
     for data in sample:
        labels, tokens = tokenizer(data['txt'])
        data['label'] = labels
        data['tokens'] = tokens
        yield data

# for ctc and whisper task, 这里逻辑不变
for batch in dataloader:
    feats, feats_lens, labels, tokens = batch

# whisper forward
def forward(...):
    ctc_tokens = tokens['ctc']
    whisper_tokens = tokens['whisper']
    ....
```

**Comment by Mddct** [comment]
我觉得 multi tokenizer先放一放，重构完简单版本tokenizer后 https://github.com/wenet-e2e/wenet/pull/2165 ，
然后重构新的dataset和IO， 到时候 这个功能就很好集成了，比如上述两种方案中的new io会更简洁

**Comment by Mddct** [comment]
> Yes，Compared with WenetTokenizer, CharTokenizer and BpeTokenizer are more clear.

先简单拆成这样的了， 先看下 https://github.com/wenet-e2e/wenet/pull/2165

**Comment by robin1001** [comment]
可以，没有问题。

**Comment by robin1001** [comment]
> 还有个需求需要考虑下，一个tokenizer允许存在俩个不同的子tokenizer。比如这个场景，ctc的词表和decoder的词表不共享，那么ctc需要一个tokenizer（如char tokenizer），decoder需要一个tokenizer（如whisper tokenizer）

除了 whisper 这个场景，这个需求还有哪些潜在场景？

**Comment by xingchensong** [comment]
> > 还有个需求需要考虑下，一个tokenizer允许存在俩个不同的子tokenizer。比如这个场景，ctc的词表和decoder的词表不共享，那么ctc需要一个tokenizer（如char tokenizer），decoder需要一个tokenizer（如whisper tokenizer）
> 
> 除了 whisper 这个场景，这个需求还有哪些潜在场景？

跟llm相关的，ctc可能都需要另一个词表。

优点：
1. llm的词表比较大
2. 为了多语言肯定对某些特定语言损失了“效率”
缺点:
1. 部署时麻烦，要两套词表

**Comment by Mddct** [comment]
@xingchensong 
https://github.com/wenet-e2e/wenet/pull/2165#issuecomment-1827532014
这个pr中提到add_special_tokens,
fintune 有些扩词表的需求， 是否需要加入 add_tokens 和模型resize_embedding

```python
tokenizer = ...
tokenizer.add_tokens("[wenet]")

model.resize_embedding() # (自动处理pretrain 参数 和 新增 （torch.cat((embed, tail), ...)）)
```

**Comment by xingchensong** [comment]
huggingface 扩充词表的issue和pr可以贴一下吗

**Comment by Mddct** [comment]
> huggingface 扩充词表的issue和pr可以贴一下吗

https://discuss.huggingface.co/t/how-to-train-the-embedding-of-special-token/10837

**Comment by xingchensong** [comment]
不过这个功能我们也是需要的（比如微调whisper，ctc的词表要加一个blank），所以同意增加这个功能

**Comment by Mddct** [comment]
tokenizer 这样设计：
1 in yaml:
```yaml
ctc_conf:
    special_tokens: <blank>

whisper_conf:
   special_tokens: <speech> # for example
```

```python
tokenizer = WhisperTokenizer()
vocab_size = tokenizer.vocab_size
tokenizer.add_special_tokens(special_tokens)
assert tokenizer.vocab.size  == vocab_size + len(special_tokens)

```

model 部分
```
model.resize_embedding(tokenizer.vocab_size)
```

**Comment by Mddct** [comment]
yaml里配置的special token，后面应该检查一下是否在tokenizer里，以及id是否对应
